README.md
# sueldoneto

A salary calculator for wages in Spain.
Concept borrowed from: 
https://cincodias.elpais.com/herramientas/calculadora-sueldo-neto/


## Usage

lein run

## License

Copyright © 2019 Adam Gefen

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-04-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-04-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/sueldoneto/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/sueldoneto/compare/0.1.0...0.1.1


test/sueldoneto/core_test.clj
(ns sueldoneto.core-test
  (:require [clojure.test :refer :all]
            [clojure.spec.alpha :as spec]
            [clojure.spec.gen.alpha :as gen]
            [clojure.spec.test.alpha :as stest]
            [sueldoneto.core :refer :all]
            [sueldoneto.coerce :refer :all]
            [sueldoneto.logic :refer :all]))

(deftest check-test
  (testing "check! resilience"
    (are [expected s v]
         (=  expected
             (coerce! s v))
      12000 :sueldoneto.core/annual-gross          "12000"
      12    :sueldoneto.core/installments          "12"
      "A"   :sueldoneto.core/personal-situation    "A"
      44    :sueldoneto.core/age                   "44"
      false :sueldoneto.core/contract              "I"
      "K"   :sueldoneto.core/professional-category "K"
      1     :sueldoneto.core/children              "1"
      1     :sueldoneto.core/young-children        "1"
      true  :sueldoneto.core/exclusivity           "S"
      2     :sueldoneto.core/ancestors             "2"
      1     :sueldoneto.core/old-ancestors         "1"
      0     :sueldoneto.core/disabled-dependents   "0"
      1     :sueldoneto.core/receiving-benefits    "1")))

(deftest check-exception-test
  (testing "check! exceptions"
    (are [expected s v]
         (= expected (try (coerce! s v)
                          (catch Exception e (.getMessage e))))
      "For input string: \"MILLION EUROS\"" :sueldoneto.core/annual-gross          "million euros"
      "For input string: \"DOCE\""          :sueldoneto.core/installments          "doce"
      "Validation failed"                   :sueldoneto.core/personal-situation    "F"
      "Validation failed"                   :sueldoneto.core/age                   "-3"
      "Validation failed"                   :sueldoneto.core/contract              "L"
      "Validation failed"                   :sueldoneto.core/professional-category "U"
      "Validation failed"                   :sueldoneto.core/children              "-1"
      "Validation failed"                   :sueldoneto.core/exclusivity           "R"
      "For input string: \"\""              :sueldoneto.core/ancestors             ""
      "For input string: \"D9\""            :sueldoneto.core/old-ancestors         "d9"
      "For input string: \"UNO\""           :sueldoneto.core/disabled-dependents   "uNo"
      "Validation failed"                   :sueldoneto.core/receiving-benefits    "-9")))

(deftest check-young-children-conditional
  (testing "coerce young-children"
    (are [expected v]
         (= expected
            (do (swap! data #(assoc % :children 2))
                (try (coerce! :sueldoneto.core/young-children v)
                     (catch Exception e (.getMessage e)))))
      1                          "1"
      "Validation failed"        "3"
      2                          "2"
      "Validation failed"        "-1"
      0                          "0"
      "For input string: \"NO\"" "no")))

;;-------------------
;;ns sueldoneto.logic
;;-------------------
(deftest max-s-social-test
  (are [expected stub]
       (= expected
          (do (swap! data #(assoc % :contract stub))
              (max-s-social)))
    (* 45014.4 0.0635) true
    (* 45014.4 0.064) false))

(deftest s-social-test
  (are [expected annual-gross contract]
       (= expected
          (do (swap! data #(assoc % :contract contract :annual-gross annual-gross))
              (s-social)))
    1280.0 20000 false
    1270.0 20000 true
    1920.0 30000 false
    1905.0 30000 true
    0.0    0     true))

(deftest disability-allowance-test
  (are [expected disability annual-gross]
       (= expected
          (do (swap! data #(assoc % :disability disability :annual-gross annual-gross))
              (disability-allowance)))
    2000       "A" 20000
    5700       "A" 10000
    2000       "A" 50000
    5500       "B" 20000
    8138.5625  "B" 13000
    5500       "B" 80000
    9200       "B" 1000
    9200       "B" 0
    13450      "C" 1000
    12388.5625 "C" 13000
    9750       "C" 30000
    9750       "C" 50000))

(deftest children-allowance-test
  (are [expected children young-children]
       (= expected
          (do (swap! data #(assoc % :children children :young-children young-children))
              (children-allowance)))
    0     0 0
    2400  1 0
    5100  2 0
    9100  3 0
    13600 4 0
    18100 5 0
    22600 6 0
    5200  1 1
    7900  2 1
    10700 2 2
    26500 5 3))

(deftest disabled-dependants-allowance-test
  (are [expected ancestors children m-grade-disabled-ancestors m-grade-disabled-descendants h-grade-disabled-ancestors h-grade-disabled-descendants]
       (= expected
          (do (swap! data #(assoc % :children children :ancestors ancestors :m-grade-disabled-ancestors m-grade-disabled-ancestors :m-grade-disabled-descendants m-grade-disabled-descendants :h-grade-disabled-ancestors h-grade-disabled-ancestors :h-grade-disabled-descendants h-grade-disabled-descendants))
              (disabled-dependants-allowance)))
    0     0 0 0 0 0 0
    0     1 1 0 0 0 0
    3000  1 1 1 0 0 0
    6000  1 1 2 0 0 0
    6000  1 1 1 1 0 0
    9000  2 1 1 2 0 0
    9000  0 1 0 0 1 0
    18000 0 2 0 0 2 0
    21000 2 1 0 1 2 0
    9000  0 1 0 0 0 1))

(comment
  (run-tests))

;;Helper in building tests
(comment
  (= (do (swap! data #(assoc % :children 1 :ancestors 2 :m-grade-disabled-ancestors 0 :m-grade-disabled-descendants 1 :h-grade-disabled-ancestors 2 :h-grade-disabled-descendants 0))
         (disabled-dependants-allowance)) 0))


src/sueldoneto/core.clj
(ns sueldoneto.core
  "Calcula el tipo de retención y el salario neto, en función de la situación laboral y familiar y del IRPF vigente."
  (:require
   [clojure.spec.alpha :as spec]
   [orchestra.core :refer [defn-spec]]
   [orchestra.spec.test :as st]
   [expound.alpha :as expound]
   [sueldoneto.messages :refer :all]
   [sueldoneto.coerce :refer :all]
   [sueldoneto.logic :refer :all])
  (:gen-class))

(spec/def ::annual-gross pos-int?)
(spec/def ::installments #{12 14})
(spec/def ::age pos-int?)
(spec/def ::personal-situation #{"A" "B" "C"})
(spec/def ::contract boolean?)
(spec/def ::professional-category #{"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K"})
(spec/def ::children  nat-int?)
(spec/def ::young-children  (spec/and #(>= (:children @data) %) nat-int?))
(spec/def ::exclusivity boolean?)
(spec/def ::ancestors nat-int?)
(spec/def ::old-ancestors nat-int?)
(spec/def ::disabled-dependents nat-int?)
(spec/def ::receiving-benefits nat-int?)
(spec/def ::m-grade-disabled-descendants (spec/and
                                          #(>= (:children @data) %)
                                          nat-int?))

(spec/def ::m-grade-disabled-ancestors (spec/and
                                        #(>= (+ (:ancestors @data) (:old-ancestors @data)) %)
                                        nat-int?))

(spec/def ::h-grade-disabled-descendants (spec/and
                                          #(>= (- (:children @data) (:m-grade-disabled-descendants @data)) %)
                                          nat-int?))
(spec/def ::h-grade-disabled-ancestors (spec/and
                                        #(>= (- (+ (:ancestors @data) (:old-ancestors @data)) (:m-grade-disabled-ancestors @data)) %)
                                        nat-int?))
(spec/def ::disability #{"A" "B" "C"})

(spec/def ::data (spec/keys :req-un
                            [::annual-gross ::installments ::personal-situation ::contract ::age ::professional-category ::children ::young-children ::exclusivity ::ancestors ::old-ancestors ::disabled-dependents ::receiving-benefits ::m-grade-disabled-descendants ::m-grade-disabled-ancestors ::h-grade-disabled-descendants ::h-grade-disabled-ancestors ::disability]))

(defn get-input
  "get user input"
  []
  (let [input (clojure.string/trim (read-line))]
    input))

(defn ppdata [db]
  (do (println (array-map "Pagas"                                      (:installments @db)
                          "\nSueldo anual: "                           (:annual-gross @db)
                          "\nSituación familiar:"                      (:personal-situation @db)
                          "\nEdad:"                                    (:age @db)
                          "\ncontrato"                                 (:contract @db)
                          "\nProfessional category:"                   (:professional-category @db)
                          "\nNiños"                                    (:children @db)
                          "\nNiños menores a 3 años"                   (:young-children @db)
                          "\nEn exclusiva"                             (:exclusivity @db)
                          "\nAscendientes"                             (:ancestors @db)
                          "\nMayores ascendientes"                     (:old-ancestors @db)
                          "\nDescapacitados"                           (:disabled-dependents @db)
                          "\nContribuyentes con minimos"               (:receiving-benefits @db)
                          "\nDescendientes discapacitados 33% y a 65%" (:m-grade-disabled-descendants @db)
                          "\nAscendientes discapacitados 33% a 65%"    (:m-grade-disabled-ancestors @db)
                          "\nDescendientes discapacitados al 65%"      (:h-grade-disabled-descendants @db)
                          "\nAscendientes discapacitados al 65%"       (:h-grade-disabled-ancestors @db)
                          "\nDiscapacidad"                             (:disability @db)
                          "\nResult-map"                               (calc)))))  ;TODO -> make a table of results

(defn prompt-disability
  []
  (println (get-in messages [:disability :pre]))
  (try
    (let [disability (coerce! ::disability (get-input))]
      (swap! data #(assoc % :disability disability))
      (ppdata data))
    (catch Exception _ (println (get-in messages [:disability :error]))
           (prompt-disability))))

(defn prompt-h-grade-disabled-ancestors
  []
  (println (get-in messages [:h-grade-disabled-ancestors :pre]))
  (try
    (let [h-grade-disabled-ancestors (coerce! ::h-grade-disabled-ancestors (get-input))]
      (swap! data #(assoc % :h-grade-disabled-ancestors h-grade-disabled-ancestors))
      (prompt-disability))
    (catch Exception e (if (= (.getMessage e) "Vlidation failed")
                         (println (get-in messages [:h-grade-disabled-ancestors :alt-error]))
                         (println (get-in messages [:h-grade-disabled-ancestors :error])))
           (prompt-h-grade-disabled-ancestors))))

(defn prompt-h-grade-disabled-descendants
  []
  (println (get-in messages [:h-grade-disabled-descendants :pre]))
  (try
    (let [h-grade-disabled-descendants (coerce! ::h-grade-disabled-descendants (get-input))]
      (swap! data #(assoc % :h-grade-disabled-descendants h-grade-disabled-descendants))
      (prompt-h-grade-disabled-ancestors))
    (catch Exception e (if (= (.getMessage e) "Validation failed")
                         (println (get-in messages [:h-grade-disabled-descendants :alt-error]))
                         (println (get-in messages [:h-grade-disabled-descendants :error])))
           (prompt-h-grade-disabled-descendants))))

(defn prompt-m-grade-disabled-ancestors
  []
  (println (get-in messages [:m-grade-disabled-ancestors :pre]))
  (try
    (let [m-grade-disabled-ancestors (coerce! ::m-grade-disabled-ancestors (get-input))]
      (swap! data #(assoc % :m-grade-disabled-ancestors m-grade-disabled-ancestors))
      (prompt-h-grade-disabled-descendants))
    (catch Exception e (if (= (.getMessage e) "Validation failed")
                         (println (get-in messages [:m-grade-disabled-ancestors :alt-error]))
                         (println (get-in messages [:m-grade-disabled-ancestors :error])))
           (prompt-m-grade-disabled-ancestors))))

(defn prompt-m-grade-disabled-descendants
  []
  (println (get-in messages [:m-grade-disabled-descendants :pre]))
  (try
    (let [m-grade-disabled-descendants (coerce! ::m-grade-disabled-descendants (get-input))]
      (swap! data #(assoc % :m-grade-disabled-descendants m-grade-disabled-descendants))
      (prompt-m-grade-disabled-ancestors))
    (catch Exception e (if (= (.getMessage e) "Validation failed")
                         (println (get-in messages [:m-grade-disabled-descendants :alt-error]))
                         (println (get-in messages [:m-grade-disabled-descendants :error])))
           (prompt-m-grade-disabled-descendants))))

(defn prompt-receiving-benefits
  []
  (println (get-in messages [:receiving-benefits :pre]))
  (try
    (let [receiving-benefits (coerce! ::receiving-benefits (get-input))]
      (swap! data #(assoc % :receiving-benefits receiving-benefits))
      (prompt-m-grade-disabled-descendants))
    (catch Exception _ (println (get-in messages [:receiving-benefits :error]))
           (prompt-receiving-benefits))))

(defn prompt-disabled-dependents
  []
  (println (get-in messages [:disabled-dependents :pre]))
  (try
    (let [disabled-dependents (coerce! ::disabled-dependents (get-input))]
      (swap! data #(assoc % :disabled-dependents disabled-dependents))
      (prompt-receiving-benefits))
    (catch Exception _ (println (get-in messages [:disabled-dependents :error]))
           (prompt-disabled-dependents))))

(defn prompt-old-ancestors
  []
  (println (get-in messages [:old-ancestors :pre]))
  (try
    (let [old-ancestors (coerce! ::old-ancestors (get-input))]
      (swap! data #(assoc % :old-ancestors old-ancestors))
      (prompt-disabled-dependents))
    (catch Exception _ (println (get-in messages [:old-ancestors :error]))
           (prompt-old-ancestors))))

(defn prompt-ancestors
  []
  (println (get-in messages [:ancestors :pre]))
  (try
    (let [ancestors (coerce! ::ancestors (get-input))]
      (swap! data #(assoc % :ancestors ancestors))
      (prompt-old-ancestors))
    (catch Exception _ (println (get-in messages [:ancestors :error]))
           (prompt-ancestors))))

(defn prompt-exclusivity
  []
  (mapv println (get-in messages [:exclusivity :pre]))
  (try
    (let [exclusivity (coerce! ::exclusivity (get-input))]
      (swap! data #(assoc % :exclusivity exclusivity))
      (prompt-ancestors))
    (catch Exception _ (println (get-in messages [:exclusivity :error]))
           (prompt-exclusivity))))

(defn prompt-young-children
  []
  (println (get-in messages [:young-children :pre]))
  (try
    (let [young-children (coerce! ::young-children (get-input))]
      (swap! data #(assoc % :young-children young-children))
      (prompt-exclusivity))
    (catch Exception e (if (= (.getMessage e) "Validation failed")
                         (println (get-in messages [:young-children :alt-error]))
                         (println (get-in messages [:young-children :error])))
           (prompt-young-children))))

(defn prompt-children
  []
  (println (get-in messages [:children :pre]))
  (try
    (let [children (coerce! ::children (get-input))]
      (swap! data #(assoc % :children children))
      (prompt-young-children))
    (catch Exception _ (println (get-in messages [:children :error]))
           (prompt-children))))

(defn prompt-professional-category
  []
  (mapv println (get-in messages [:professional-category :pre]))
  (try
    (let [professional-category (coerce! ::professional-category (get-input))]
      (swap! data #(assoc % :professional-category professional-category))
      (prompt-children))
    (catch Exception _ (println (get-in messages [:professional-category :error]))
           (prompt-professional-category))))

(defn prompt-contract
  []
  (mapv println (get-in messages [:contract :pre]))
  (try
    (let [contract (coerce! ::contract (get-input))]
      (swap! data #(assoc % :contract contract))
      (prompt-professional-category))
    (catch Exception _ (println (get-in messages [:contract :error]))
           (prompt-contract))))

(defn prompt-age
  []
  (println (get-in messages [:age :pre]))
  (try
    (let [age (coerce! ::age (get-input))]
      (swap! data #(assoc % :age age))
      (prompt-contract))
    (catch Exception _ (println (get-in messages [:age :error]))
           (prompt-age))))

(defn prompt-personal-situation
  []
  (mapv println (get-in messages [:personal-situation :pre]))
  (try
    (let [personal-situation (coerce! ::personal-situation (get-input))]
      (swap! data #(assoc % :personal-situation personal-situation))
      (prompt-age))
    (catch Exception _ (println (get-in messages [:personal-situation :error]))
           (prompt-personal-situation))))

(defn prompt-installments
  []
  (println (get-in messages [:installments :pre]))
  (try
    (let [installments (coerce! ::installments (get-input))]
      (swap! data #(assoc % :installments installments))
      (prompt-personal-situation))
    (catch Exception _ (println (get-in messages [:installments :error]))
           (prompt-installments))))

(defn prompt-annual-gross
  []
  (println (get-in messages [:annual-gross :pre]))
  (try
    (let [annual-gross (coerce! ::annual-gross (get-input))]
      (swap! data #(assoc % :annual-gross annual-gross))
      (prompt-installments))
    (catch Exception _ (println (get-in messages [:annual-gross :error]))
           (prompt-annual-gross))))

(defn start-app []
  (println "Calculadora de sueldo neto")
  (println "''''''''''''''''''''''''''")
  (println "")
  (prompt-annual-gross))

(defn -main
  [& args]
  (start-app))


src/sueldoneto/coerce.clj
(ns sueldoneto.coerce
  (:require
   [clojure.spec.alpha :as spec]))

(def built-in-coercions
  {`:sueldoneto.core/annual-gross                 #(Integer/parseInt %)
   `:sueldoneto.core/installments                 #(Integer/parseInt %)
   `:sueldoneto.core/age                          #(Integer/parseInt %)
   `:sueldoneto.core/contract                     #(cond (= % "G") true
                                                         (= % "I") false)
   `:sueldoneto.core/children                     #(Integer/parseInt %)
   `:sueldoneto.core/young-children               #(Integer/parseInt %)
   `:sueldoneto.core/exclusivity                  #(cond (= % "S") true
                                                         (= % "N") false)
   `:sueldoneto.core/ancestors                    #(Integer/parseInt %)
   `:sueldoneto.core/old-ancestors                #(Integer/parseInt %)
   `:sueldoneto.core/disabled-dependents          #(Integer/parseInt %)
   `:sueldoneto.core/receiving-benefits           #(Integer/parseInt %)
   `:sueldoneto.core/m-grade-disabled-descendants #(Integer/parseInt %)
   `:sueldoneto.core/m-grade-disabled-ancestors   #(Integer/parseInt %)
   `:sueldoneto.core/h-grade-disabled-descendants #(Integer/parseInt %)
   `:sueldoneto.core/h-grade-disabled-ancestors   #(Integer/parseInt %)})

(defn coerce [key value]
  (let [coerce-fn (get built-in-coercions key identity)]
    (if (string? value)
      (coerce-fn (clojure.string/upper-case value))
      value)))

(defn check! [spec value]
  (or (spec/valid? spec value)
      (throw (ex-info "Validation failed" {:explanation (spec/explain-str spec value)}))))

(defn coerce!
  [spec v]
  (let [v (coerce spec v)]
    (check! spec v)
    v))



src/sueldoneto/messages.clj
(ns sueldoneto.messages)
(def messages
  {:annual-gross                 {:pre   "Sueldo bruto anual:"
                                  :error "Accepta solo -> Euros en números"}
   :installments                 {:pre   "Número de pagas:"
                                  :error "Accepta solo -> 12 o 14"}
   :age                          {:pre   "Edad:"
                                  :error "Accepta solo -> Años en números"}
   :personal-situation           {:pre   ["Situación familiar:"
                                          "[A] Soltero, viudo, divorciado o separado con hijos a cargo"
                                          "[B] Casado y cuyo cónyuge no obtiene rentas superiores a 1.500 euros anuales"
                                          "[C] Otros"]
                                  :error "Accepta solo -> \"A\" \"B\" o \"C\""}
   :contract                     {:pre   ["Tipo de contrato laboral:"
                                          "[G] general [I] Duración inferior a doce meses"]
                                  :error "Accepta solo \"G\" o \"I\""}
   :professional-category        {:pre   ["Categoría profesional:"
                                          "[A] Ingenieros y Licenciados"
                                          "[B] Ingenieros Técnicos, Peritos y Ayudantes Titulados"
                                          "[C] Jefes Administrativos y de Taller"
                                          "[D] Ayudantes no Titulados"
                                          "[E] Oficiales Administrativos"
                                          "[F] Subalternos"
                                          "[G] Auxiliares Administrativos"
                                          "[H] Oficiales de primera y segunda"
                                          "[I] Oficiales de tercera y Especialistas"
                                          "[J] Peones"
                                          "[K] Trabajadores menores de dieciocho años, cualquiera"]
                                  :error "Accepta solo -> \"A\" \"B\" \"C\" \"D\" \"E\" \"F\" \"G\" \"H\" \"I\" \"J\" o \"K\""}
   :children                     {:pre   "Número de hijos menores de 25 años:"
                                  :error "Accepta solo -> Cantidad en números"}
   :young-children               {:pre       "De sus hijos, cuántos tienen menos de 3 años?"
                                  :error     "Accepta solo -> Cantidad en números"
                                  :alt-error "Accepta catindad igual o inferior al campo anterior"}
   :exclusivity                  {:pre   ["Tiene los hijos en exclusiva a efectos fiscales?" "[S] si" "[N] no"]
                                  :error "Accepta solo -> \"s\" o \"n\""}
   :ancestors                    {:pre   "Mayores de 65 años y menores de 75 años a cargo"
                                  :error "Accepta solo -> Cantidad en números"}
   :old-ancestors                {:pre   "Número de ascendientes mayores de 75 años a cargo"
                                  :error "Accepta solo -> Cantidad en números"}
   :disabled-dependents          {:pre   "Menor de 65 años a cargo con discapacidad"
                                  :error "Accepta solo -> Cantidad en números"}
   :receiving-benefits           {:pre   "Contribuyentes que aplican los mínimos por ascendiente"
                                  :error "Accepta solo -> Cantidad en números"}
   :m-grade-disabled-descendants {:pre       "Número de descendientes con grado de discapacidad entre el 33% y el 65%"
                                  :error     "Accepta solo -> Cantidad en números"
                                  :alt-error "Accepta cantindad igual o inferior al cantidad de niños con menos de 25 años"}
   :m-grade-disabled-ancestors   {:pre       "Número de ascendientes con grado de discapacidad entre el 33% y el 65%"
                                  :error     "Accepta solo -> Cantidad en números"
                                  :alt-error "Accepta cantindad igual o inferior al cantidad de ascendientes mayores que 65 años a cargo"}
   :h-grade-disabled-descendants {:pre       "Número de descendientes con grado de discapacidad igual o superior al 65%"
                                  :error     "Accepta solo -> Cantidad en números"
                                  :alt-error "Accepta cantindad igual o inferior al cantidad de niños con menos de 25 años menos los con grado de discapacidad entre el 33% y el 65%"}
   :h-grade-disabled-ancestors   {:pre       "Número de ascendientes con grado de discapacidad igual o superior al 65%"
                                  :error     "Accepta solo -> Cantidad en números"
                                  :alt-error "Accepta cantidad igual o inferior al cantidad de ascendientes mayores que 65 años a cargo menos los con grado de discapacidad entre el 33% y el 65%"}
   :disability {:pre ["Disacpacidad (del contribuyente):" "[A] no tiene ninguna descapacidad" "[B] Grado de discapacidad entre el 33% y el 65%" "[C] Grado de discapacidad igual o superior al 65% o con movilidad reducida"]
                :error "Accepta solo -> \"A\" \"B\" o \"C\""}})



src/sueldoneto/logic.clj
(ns sueldoneto.logic
  (:require
   [sueldoneto.messages :refer [messages]]
   [sueldoneto.coerce :refer :all]))

(def data (atom nil))

;;stub -> uncomment in development

(comment
  (swap! data #(assoc % :annual-gross 30000 :installments 12 :disability "B" :contract true :professional-category "C" :children 3 :young-children 1 :ancestors 1 :old-ancestors 1 :m-grade-disabled-ancestors 0 :m-grade-disabled-descendants 0 :h-grade-disabled-ancestors 0 :h-grade-disabled-descendants 1)))

;;The cap for annual income's Seguridad Social 6.4% contribution, 

(def max-gross 45014.4)  ;Euros
(def general-contract-rate 0.0635)
(def inferior-contract-rate 0.064)

(defn max-s-social []
  (let [{:keys [contract]} @data]
    (-> max-gross
        (* (if contract
             general-contract-rate
             inferior-contract-rate)))))

(comment ;TODO find a more interesting implemntation of this
  (defn cuota-mensual-pagar []
    (let [{:keys [annual-gross professional-category]} @data
          datos (case professional-category
                  "A" {:min 1052.9 :max 3751.2}
                  "B" {:min 956.1 :max 3751.2}
                  "C" {:min 831.6 :max 3751.2}
                  "D" {:min 825.6 :max 3751.2}
                  "E" {:min 825.6 :max 3751.2}
                  "F" {:min 825.6 :max 3751.2}
                  "G" {:min 825.6 :max 3751.2}
                  "H" {:min 825.6 :max 3751.2}
                  "I" {:min 825.6 :max 3751.2}
                  "J" {:min 825.6 :max 3751.2}
                  "K" {:min 825.6 :max 3751.2})]
      (if (< (/ annual-gross 12) (:min datos))
        (* 0.0635 (:min datos))
        (if (> (/ annual-gross 12) (:max datos))
          (* 0.0635 (:min datos))
          (* (/ annual-gross 12) 0.0635))))))

(defn s-social
  []
  (let [{:keys [annual-gross contract]} @data
        percentage  (-> annual-gross
                        (* (if contract
                             general-contract-rate
                             inferior-contract-rate)))]
    (if (< percentage (max-s-social))
      percentage
      (max-s-social))))

(defn disability-allowance []
  (let [{:keys [disability annual-gross]} @data
        net-profit (- annual-gross (s-social))
        base-exemption 2000
        net-reduction (cond
                        (< net-profit 11250) 3700
                        (>= net-profit 14450) 0
                        :else (- 3700 (* 1.15625 (- net-profit 11250))))
        disability-reduction (case disability
                               "A"  0
                               "B"  3500
                               "C"  7750)]
    (+ base-exemption net-reduction disability-reduction)))

(defn children-allowance
  []
  (let [{:keys [children young-children]} @data
        extra-children (- children 4)
        allowance (case children
                    0 0
                    1 2400
                    2 5100
                    3 9100
                    4 13600
                    (+ 13600 (* 4500 extra-children)))]
    (+ allowance (* 2800 young-children))))

(defn ancestor-allowance
  []
  (let [{:keys [annual-gross ancestors old-ancestors]} @data
        net-profit (- annual-gross (s-social))]
    (if (< net-profit 8000)
      (+ (* ancestors 1150) (* old-ancestors 2550))
      0)))

(defn disabled-dependants-allowance
  []
  (let [{:keys [m-grade-disabled-descendants m-grade-disabled-ancestors h-grade-disabled-descendants h-grade-disabled-ancestors]} @data
        m-grade (+ m-grade-disabled-ancestors m-grade-disabled-descendants)
        h-grade (+ h-grade-disabled-ancestors h-grade-disabled-descendants)]
    (+ (* m-grade 3000) (* h-grade 9000))))

(defn total-allowance
  []
  (+ (ancestor-allowance) (children-allowance) (disability-allowance) (disabled-dependants-allowance)))

(defn family-situation-exemption
  []
  (let [{:keys [children personal-situation]} @data]
    (cond
      (= "B" personal-situation) (case children
                                   0 13696
                                   1 14985
                                   17138)
      (= "C" personal-situation) (case children
                                   0 12000
                                   1 12607
                                   13275)
      :else (case children
              0 0
              1 14266
              15803))))

(defn irpf
  []
  (let [{:keys [annual-gross]} @data
        net-profit (- annual-gross (s-social))
        base (- net-profit (total-allowance))
        a (if (<= base 12450) base 12450)
        b (if (<= base 20200) 0 (- base 12450))
        c (if (<= base 35200) 0 (- base 20200))
        d (if (<= base 60000) 0 (- base 35200))
        e (if (> base 60000) (- base 60000) 0)]
    (if (< net-profit (family-situation-exemption))
      0
      (+ (* a 0.19) (* b 0.24) (* c 0.3) (* d 0.37) (* e 0.45)))))

(defn calc
  []
  (let [{:keys [annual-gross installments contract]} @data
        s-social-contribution (s-social)
        irpf (irpf)
        annual-net (-> annual-gross
                       (- s-social-contribution irpf))]
    {:net-wage (/ annual-net installments) :s-social-contribution (/  s-social-contribution installments) :irpf (/ irpf installments)}))

